### PR TITLE
fix(treesitter): properly restore `'syntax'`

### PR DESCRIPTION
**Pull request by lewis6991**

### CODE REVIEW

**Review by clason — APPROVED** [review]

**Comment by clason** [comment]
Not tested, but that seems the right place to handle this. (Long-term, we of course want to integrate treesitter into `syntax` so these shenanigans are no longer needed.)

**Comment by lewis6991** [comment]
asan failure on a lua-only change ....spicy.

**Comment by zeertzjq** [comment]
It's buffer-updates unload callback again

File: src/nvim/memory.c [hunk 1]
```diff
@@ -812,6 +813,11 @@ void free_all_mem(void)
     bufref_T bufref;
     set_bufref(&bufref, buf);
     nextbuf = buf->b_next;
+
+    // Since options have been freed above we need to ensure no callbacks are called,
```
Comment (zeertzjq) [inline src/nvim/memory.c:817]: This comment may be a bit strange. A lot of stuff have been freed above, not just options.

Comment (zeertzjq) [reply src/nvim/memory.c]: Other being freed can cause other issues.

Comment (lewis6991) [reply src/nvim/memory.c]: I've updated the comment.

**Review by bfredl — APPROVED** [review]

**Pull request merged**